core/src/trezor/ui/layouts/tr/__init__.py

```diff
@@ -59,232 +205,1069 @@ def handle_timers(self) -> loop.Task:  # type: ignore [awaitable-is-generator]
             msg = self.layout.timer(token)
             if msg is not None:
                 raise ui.Result(msg)
-            self.layout.paint()
-            ui.refresh()
+            self._paint()
 
+    def _place_layout(self) -> None:
+        """It is necessary to place the layout to get data about its screen content."""
+        self.layout.place()
 
-async def confirm_action(
-    ctx: wire.GenericContext,
+    def _read_content_raw(self) -> list[str]:
+        """Reads raw trace content from Rust layout."""
+        result: list[str] = []
+
+        def callback(*args: str):
+            for arg in args:
+                result.append(str(arg))
+
+        self.layout.trace(callback)
+        return result
+
+    def page_count(self) -> int:
+        """How many paginated pages current screen has."""
+        # TODO: We could do it some other way
+        # TT does this:
```

Comment:
Yeah, why not do it the same way?

Reply:
Done like that in [ef1edc5](https://github.com/trezor/trezor-firmware/pull/2610/commits/ef1edc5995619af91746b05b04f2ec2ff13e08ca)